### PR TITLE
feat(log manager): skip filtering by date when a static path is used

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
Currently the log manager is filtering logs by date no matter whether provided path is static or containing wildcard. With this PR log manager will return log path if it is fixed. Explanation why this feature is needed can be found in #2966.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
* #2966 
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/log_manager/src/log_utils.rs [hunk 1]
```diff
@@ -93,68 +93,65 @@ pub fn read_log_content(
     }
 }
 
-pub fn filter_logs_on_type(
+pub fn filter_logs(
```
Comment (didier-wenzek) [inline crates/common/log_manager/src/log_utils.rs:96]: This function does what is expected by [#2966](https://github.com/thin-edge/thin-edge.io/issues/2966), but is a bit convoluted.

I would try to use a pipeline style:
- generate a list of `(path, modification_date, wildcard_match)`
- filter out paths out of date range (unless not being the result of a wildcard match)
- sort by modification date
- map to keep only the paths

File: crates/common/log_manager/src/log_utils.rs [hunk 2]
```diff
@@ -93,42 +93,73 @@ pub fn read_log_content(
...
-) -> Result<Vec<PathBuf>, LogRetrievalError> {
-    let mut out = vec![];
+) -> Vec<PathBuf> {
+    // include log file with static path no matter whether or not it is in date range
+    files
+        .into_iter()
+        .filter(|(_, modification_time, wildcard_match)| {
+            !wildcard_match || *modification_time >= date_from
+        })
+        .map(|(path, _, _)| path)
```
Comment (didier-wenzek) [inline crates/common/log_manager/src/log_utils.rs:157]: I would do the sort by modification date here, before discarding the modification date that has already been extracted.

**Comment by didier-wenzek** [comment]
> Question: when we are filtering logs by date, why only lower limit is taken into account (`date_from`) and we don't check upper limit (`date_to`)?

I think thin-edge uses these two time-related filters in an unexpected way.  The intent is to filter the log files using the timestamps of the log entries. However, thin-edge doesn't even try to decode these timestamps and only guesses that some entries are more recent that the `date_from` using the file modification date. Using `date_to` the same way would be really confusing (a whole log being discarded simply because of a recent entry).

**Comment by rina23q** [comment]
> > Question: when we are filtering logs by date, why only lower limit is taken into account (`date_from`) and we don't check upper limit (`date_to`)?
> 
> I think thin-edge uses these two time-related filters in an unexpected way. The intent is to filter the log files using the timestamps of the log entries. However, thin-edge doesn't even try to decode these timestamps and only guesses that some entries are more recent that the `date_from` using the file modification date. Using `date_to` the same way would be really confusing (a whole log being discarded simply because of a recent entry).

It was a difficult decision how to "interpret" the c8y's keys `DateTo` and `DateFrom` to thin-edge. For some log types like syslog, there are timestamps in each line. If all logs were like that, the design would have been different (filter `inside` log files)

On the other hand, thin-edge needs a generic solution. Then, we are using `DateFrom` to filter by the file's modification time. I think it's an oversight that we don't use `DateTo` to filter that. However, I don't see the value to be very strict there since our solution is anyway not very ideal as I mentioned above.

File: crates/common/log_manager/src/log_utils.rs [hunk 3]
```diff
@@ -172,95 +185,114 @@ mod tests {
...
+                config_type: "type_one".to_string(),
+            },
+            FileEntry {
+                path: format!("{tempdir_path}/file_c_two"),
+                config_type: "type_two".to_string(),
+            },
+            FileEntry {
+                path: format!("{tempdir_path}/file_d_one"),
+                config_type: "type_one".to_string(),
+            },
```
Comment (rina23q) [inline crates/common/log_manager/src/log_utils.rs:285]: I'm actually surprised that user can define multiple entries with the same `config_type`. But it makes sense. Good test case.

**Review by rina23q — COMMENTED** [review]
I think it would be better to modify one system test.

Just giving out of range timestamp to `${start_timestamp}` in this test. Then the test will cover your change. Probably renaming test name makes also sense so that it tells that we ignore `dateFrom` and `dateTo` for a static path.

https://github.com/thin-edge/thin-edge.io/blob/4fbf29f913643c0df8f2609caa6da749f54d0d2d/tests/RobotFramework/tests/cumulocity/log/log_operation.robot#L16-L23

For the wildcard case, the `Log file request supports date/time filters and can search across multiple log files` test already covered enough, so no action needed.

**Review by gligorisaev — COMMENTED** [review]
The PR addresses the issues highlighted in #2966

### Testing Performed

#### Integration Tests:

- Conducted integration tests to validate the end-to-end functionality of the error handling improvements.
- Simulated various MQTT connection error scenarios to verify the effectiveness of the new logic.

#### Manual Testing:

- Manually triggered MQTT connection errors to observe the system's behavior.
- Verified the accuracy and helpfulness of the error messages produced.

### Test Results

#### Manual Testing: 

- Manual tests corroborated the automated test results, demonstrating clear and actionable error messages during connection issues.

#### Observations
- The improved error handling significantly enhances the reliability of the tedge-mapper.
- The addition of retry logic helps mitigate transient connection issues without manual intervention.

**The implementation is robust, thoroughly tested, and ready for deployment.**

**Review by rina23q — APPROVED** [review]
LGTM

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**